src/IECore/MeshAlgo.cpp

```diff
@@ -0,0 +1,202 @@
+
+	std::vector<V3f> uTangents( numUniqueTangents, V3f( 0 ) );
+	std::vector<V3f> vTangents( numUniqueTangents, V3f( 0 ) );
+	std::vector<V3f> normals( numUniqueTangents, V3f( 0 ) );
+
+	for( size_t faceIndex = 0; faceIndex < vertsPerFace.size(); faceIndex++ )
+	{
+		assert( vertsPerFace[faceIndex] == 3 );
+
+		// indices into the facevarying data for this face
+		size_t fvi0 = faceIndex * 3;
+		size_t fvi1 = fvi0 + 1;
+		size_t fvi2 = fvi1 + 1;
+		assert( fvi2 < vertIds.size() );
+		assert( fvi2 < u.size() );
+		assert( fvi2 < v.size() );
+
+		// positions for each vertex of this face
+		const V3f &p0 = points[vertIds[fvi0]];
+		const V3f &p1 = points[vertIds[fvi1]];
+		const V3f &p2 = points[vertIds[fvi2]];
+
+		// uv coordinates for each vertex of this face
+		const V2f uv0( u[fvi0], v[fvi0] );
+		const V2f uv1( u[fvi1], v[fvi1] );
+		const V2f uv2( u[fvi2], v[fvi2] );
+
+		// compute tangents and normal for this face
+		const V3f e0 = p1 - p0;
+		const V3f e1 = p2 - p0;
+
+		const V2f e0uv = uv1 - uv0;
+		const V2f e1uv = uv2 - uv0;
+
+		V3f tangent = ( e0 * -e1uv.y + e1 * e0uv.y ).normalized();
+		V3f bitangent = ( e0 * -e1uv.x + e1 * e0uv.x ).normalized();
+
+		V3f normal = ( p2 - p1 ).cross( p0 - p1 );
+		normal.normalize();
+
+		// and accumlate them into the computation so far
+		uTangents[stIndices[fvi0]] += tangent;
+		uTangents[stIndices[fvi1]] += tangent;
+		uTangents[stIndices[fvi2]] += tangent;
+
+		vTangents[stIndices[fvi0]] += bitangent;
+		vTangents[stIndices[fvi1]] += bitangent;
+		vTangents[stIndices[fvi2]] += bitangent;
+
+		normals[stIndices[fvi0]] += normal;
+		normals[stIndices[fvi1]] += normal;
+		normals[stIndices[fvi2]] += normal;
+
+	}
+
+	// normalize and orthogonalize everything
+	for( size_t i = 0; i < uTangents.size(); i++ )
+	{
+		normals[i].normalize();
+
+		uTangents[i].normalize();
+		vTangents[i].normalize();
+
+		// Make uTangent/vTangent orthogonal to normal
+		uTangents[i] -= normals[i] * uTangents[i].dot( normals[i] );
+		vTangents[i] -= normals[i] * vTangents[i].dot( normals[i] );
+
+		uTangents[i].normalize();
+		vTangents[i].normalize();
+
+		if( orthoTangents )
+		{
+			vTangents[i] -= uTangents[i] * vTangents[i].dot( uTangents[i] );
+			vTangents[i].normalize();
+		}
+
+		// make things less sinister
+		if( uTangents[i].cross( vTangents[i] ).dot( normals[i] ) < 0.0f )
+		{
+			uTangents[i] *= -1.0f;
+		}
+	}
+
+	// convert the tangents back to facevarying data and add that to the mesh
+	V3fVectorDataPtr fvUD = new V3fVectorData();
+	V3fVectorDataPtr fvVD = new V3fVectorData();
+
+	std::vector<V3f> &fvU = fvUD->writable();
+	std::vector<V3f> &fvV = fvVD->writable();
+	fvU.resize( stIndices.size() );
+	fvV.resize( stIndices.size() );
+
+	for( unsigned i = 0; i < stIndices.size(); i++ )
+	{
+		fvU[i] = uTangents[stIndices[i]];
+		fvV[i] = vTangents[stIndices[i]];
+	}
+
+	PrimitiveVariable::Interpolation tangentInterpolation = mesh->inferInterpolation( stIndicesData );
```

Comment:
`Primitive::inferInterpolation()` is a bit of a last resort, because there are ambiguous cases. In the original version this was always FaceVarying, because we knew that was the only thing we were supporting.